### ISSUE TITLE
parser error with "".empty?? true : false 0.10.0

### ISSUE THREAD

**Issue by cj** [opal/opal#1428]
running:

`$ bundle exec opal -e '"".empty?? true : false'`

gives:

```
/home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/parser.rb:82:in `on_error': An error occurred while compiling: -e (RuntimeError)

parse error on value ":" (tCOLON) :-e:1 Source: -e:1:17
"".empty?? true : false
~~~~~~~~~~~~~~~~^
from /home/cj/.rbenv/versions/2.2.4/lib/ruby/2.2.0/racc/parser.rb:258:in `_racc_do_parse_c'
from /home/cj/.rbenv/versions/2.2.4/lib/ruby/2.2.0/racc/parser.rb:258:in `do_parse'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/parser.rb:54:in `parse_to_sexp'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/parser.rb:39:in `parse'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/compiler.rb:151:in `compile'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/builder_processors.rb:91:in `compiled'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/builder_processors.rb:101:in `requires'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/builder.rb:41:in `build_str'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/cli.rb:111:in `block in build'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/cli.rb:156:in `evals_or_file'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/cli.rb:110:in `build'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/cli.rb:121:in `compiled_source'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/cli.rb:85:in `run_code'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/lib/opal/cli.rb:71:in `run'
from /home/cj/apps/acd/autolink/.bundle/gems/opal-0.10.0.beta2/bin/opal:22:in `<top (required)>'
from /home/cj/apps/acd/autolink/.bundle/bin/opal:16:in `load'
from /home/cj/apps/acd/autolink/.bundle/bin/opal:16:in `<main>'
```

**Comment by iliabylich** [comment]
This issue is a bit tricky. In MRI `empty??` is a valid method name:
``` ruby
define_method("empty??") { 1 }
send("empty??")
# => 1
```

And that's why opal thinks that it's a valid identifier:
```
bundle exec opal -c --no-opal -e '"".empty?? true : false'
[:tSTRING_BEG, "\""]
[:tSTRING_END, "\""]
[:tDOT, "."]
[:tIDENTIFIER, "empty??"]
[:kTRUE, "true"]
[:tCOLON, ":"]
...error...
```

But of course, identifier != method_name. These two constants should be extended - https://github.com/opal/opal/blob/master/lib/opal/regexp_anchors.rb#L15-L16 

By the way, MRI allows you to define even more weird methods:

``` ruby
define_method("m!@#$%^&*-=_+") { 1 }
send("m!@#$%^&*-=_+")
# => Yes, it works :D
```

**Comment by cj** [comment]
@iliabylich the way I run into the issue was similar to this (which I know lots of people do).

```ruby
def is_admin?
   current_user.role.admin
end

is_admin?? "do some admin thingy" : "you're not an admin!"
```

p.s. this worked in earlier versions of opal.

**Comment by cj** [comment]
`'0.8.0'` is the version I know it worked in.

**Comment by iliabylich** [comment]
You are 100% correct, that's a regression caused by https://github.com/opal/opal/commit/5716c3456feabe2280091b8f623f147ff109d789. I'll take a look at this issue tomorrow.

**Comment by cj** [comment]
@iliabylich there is still some regression.  I'm now getting this error in my code which I didn't in `0.8.0`.

```
parse error on value "do" (kDO) :components/dashboard/events:148
Source: components/dashboard/events:148:41
                            on :click, '.portlet td.watch i' do
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^
/home/cj/apps/acd/autolink/.bundle/bundler/gems/opal-e10eb231b647/lib/opal/parser.rb:82:in `on_error'
```

**Comment by iliabylich** [comment]
`opal -c --no-opal -e "on :click, '.portlet td.watch i' do; end"`

```
/* Generated by Opal 0.10.0.beta2 */
(function(Opal) {
  var $a, $b, TMP_1, self = Opal.top, $scope = Opal, nil = Opal.nil, $breaker = Opal.breaker, $slice = Opal.slice;

  Opal.add_stubs(['$on']);
  return ($a = ($b = self).$on, $a.$$p = (TMP_1 = function(){var self = TMP_1.$$s || this;

  return nil}, TMP_1.$$s = self, TMP_1.$$arity = 0, TMP_1), $a).call($b, "click", ".portlet td.watch i")
})(Opal);

/* Generated by Opal 0.10.0.beta2 */
(function(Opal) {
  var self = Opal.top, $scope = Opal, nil = Opal.nil, $breaker = Opal.breaker, $slice = Opal.slice;

  Opal.add_stubs(['$exit']);
  return $scope.get('Kernel').$exit()
})(Opal);
```

@cj Are you sure that you are on the latest opal master? I can compile this code locally on e10eb231b6472394ff46d9af55bca14cae49fc9c (and compiled code seems to be correct)

**Comment by cj** [comment]
Gemfile.lock says `revision: e10eb231b6472394ff46d9af55bca14cae49fc9c `.  It only throws that error with the entire file, there are other lines in the file exactly the same as that one and it doesn't throw an error.

**Comment by cj** [comment]
https://gist.github.com/862f9d12ca5b126e7357d3705ba7a1bd

**Comment by cj** [comment]
line 148

**Comment by cj** [comment]
if I change it to `on(:click, '.portlet td.watch i') do`, I no longer get the compile error.

**Comment by iliabylich** [comment]
@cj the issue was fixed in #1438
